File: app/models/tournament.rb
class Tournament < ActiveRecord::Base
  has_one :bracket
  belongs_to :user
  has_many :competitorships
  has_many :users, through: :competitorships
  has_many :teams, through: :competitorships

  def start
    self.started = true
    bracket = Bracket.create
    bracket.build(participants: participants)
    self.bracket = bracket
    self.save
  end

  def clear
    self.started = false
    self.bracket = nil
    self.save
  end

  def participants
    self.users.present? ? self.users : self.teams
  end
end

File: app/assets/javascripts/models/tournament.js
attr = DS.attr;

Gameway.Tournament = DS.Model.extend({
  name: attr('string'),
  bracketId: attr('string'),
  // bracket: function() {
  //   this.store.find('bracket', this.get('bracketId')).then(function(bracket) {
  //     return bracket
  //   }, function(er) { debugger; });
  // }.property('bracketId'),
  description: attr('string'),
  user: DS.belongsTo('user'),
  teams: DS.hasMany('team'),
  started: attr('boolean'),
  startsAt: attr('date'),
  ended: attr('boolean'),
  lolRegion: attr('string'),
  createdAt: attr('date'),
  updatedAt: attr('date'),
  start: function() {
    var bracket = this.store.createRecord('bracket');
    this.set('started', true);
    bracket.build({ participants: this.get('participants'), tournament: this });
    this.set('bracketId', bracket.get('id'));
    this.save();
  },
  participants: function() {
    if (Ember.isNone(this.get('users'))) {
      return this.get('teams');
    } else {
      return this.get('users');
    }
  }.property('teams', 'users')
})
